src/abstractdatatable/join.jl

```diff
@@ -44,50 +44,66 @@ Base.length(x::RowIndexMap) = length(x.orig)
         @inbounds all_orig_left_ixs[leftonly_ixs.join] = leftonly_ixs.orig
     else
         # the result contains only the left rows that are matched to right rows (left_ixs)
         all_orig_left_ixs = left_ixs.orig # no need to copy left_ixs.orig as it's not used elsewhere
     end
-    ril = length(right_ixs)
-    loil = length(leftonly_ixs)
-    roil = length(rightonly_ixs)
-    left_dt = DataTable(Any[resize!(col[all_orig_left_ixs], length(all_orig_left_ixs)+roil)
-                            for col in columns(joiner.dtl)],
-                        names(joiner.dtl))
-
-    # compose right half of the result taking all right columns excluding on
-    dtr_noon = without(joiner.dtr, joiner.on_cols)
     # permutation to swap rightonly and leftonly rows
     right_perm = vcat(1:ril, ril+roil+1:ril+roil+loil, ril+1:ril+roil)
     if length(leftonly_ixs) > 0
         # compose right_perm with the permutation that restores left rows order
         right_perm[vcat(right_ixs.join, leftonly_ixs.join)] = right_perm[1:ril+loil]
     end
     all_orig_right_ixs = vcat(right_ixs.orig, rightonly_ixs.orig)
-    right_dt = DataTable(Any[resize!(col[all_orig_right_ixs], length(all_orig_right_ixs)+loil)[right_perm]
-                             for col in columns(dtr_noon)],
-                         names(dtr_noon))
-    # merge left and right parts of the joined table
-    res = hcat!(left_dt, right_dt)
+
+    # compose right half of the result taking all right columns excluding on
+    dtr_noon = without(joiner.dtr, joiner.on_cols)
+
+    nrow = length(all_orig_left_ixs) + roil
+    @assert nrow == length(all_orig_right_ixs) + loil
+    ncleft = ncol(joiner.dtl)
+    cols = Vector{Any}(ncleft + ncol(dtr_noon))
+    _similar_type = kind == :inner ? similar : similar_nullable
+    for (i, col) in enumerate(columns(joiner.dtl))
+        cols[i] = _similar_type(col, nrow)
+        @inbounds for (j, k) in enumerate(all_orig_left_ixs)
+            cols[i][j] = col[k]
+        end
+    end
+    for (i, col) in enumerate(columns(dtr_noon))
+        cols[i+ncleft] = _similar_type(col, nrow)
+        @inbounds for (j, k) in enumerate(all_orig_right_ixs)
+            cols[i+ncleft][j] = col[k]
+        end
+        permute!(cols[i+ncleft], right_perm)
+    end
+    res = DataTable(cols, vcat(names(joiner.dtl), names(dtr_noon)))
 
     if length(rightonly_ixs.join) > 0
         # some left rows are nulls, so the values of the "on" columns
         # need to be taken from the right
         for (on_col_ix, on_col) in enumerate(joiner.on_cols)
             # fix the result of the rightjoin by taking the nonnull values from the right table
-            res[on_col][rightonly_ixs.join] = joiner.dtr_on[rightonly_ixs.orig, on_col_ix]
+            # end-length(rightonly_ixs.orig)+1:end was rightonly_ixs.join. Try and FIXME
+            res[on_col][end-length(rightonly_ixs.orig)+1:end] = joiner.dtr_on[rightonly_ixs.orig, on_col_ix]
```

Comment:
The FIXME refers to `rightonly_ixs.join` and that I would like to change this line back to using `res[on_col][rightonly_ixs.join] = ...` when we get around to figuring out why `rightonly_ixs.join` doesn't contain the expected values. I could remove the FIXME and open an issue? I could also remove the FIXME and just keep a personal note to get around to fixing this?  Not sure what the preferred way of tracking this would be.